lib/task.js

```diff
@@ -440,40 +565,74 @@ export class Task {
   }
 
   /**
-  Determine when the right time is to issue another reclaim then schedule it
-  via set timeout.
-  */
-  scheduleReclaim(claim) {
-    // Figure out when to issue the next claim...
-    let takenUntil = (new Date(claim.takenUntil) - new Date());
-    let nextClaim = takenUntil / this.runtime.task.reclaimDivisor;
+   * Resolves all of the non-primary claims for this task, optionally
+   * adding an artifact to each one named "public/superseded-by"
+   * containing the taskId/runId of the primary claim.
+   *
+   * @param {Integer} primaryTaskId taskId of the primary task
+   * @param {Integer} primaryRunId runId of the primary task
+   * @param {Boolean} addArtifacts if true, add the superseded-by artifacts
+   * @param {String} reason the exception reason passed to the queue
+   */
 
-    // This is tricky ensure we have logs...
-    this.runtime.log('next claim', {
-      taskId: this.status.taskId,
-      runId: this.runId,
-      time: nextClaim
-    });
+  async resolveSuperseded(primaryTaskId, primaryRunId, addArtifacts, reason) {
+    let queue = this.runtime.queue;
+    let log = this.runtime.log;
+    let supersedes = [];
 
-    // Figure out when we need to make the next claim...
-    this.clearClaimTimeout();
+    try {
+      await Promise.all(this.claims.map(async function(c){
+        let taskId = c.status.taskId;
+        let runId = c.runId;
+        if (taskId == primaryTaskId && runId == primaryRunId) {
+          return;
+        }
 
-    this.claimTimeoutId =
-      setTimeout(function() {
-        async () => {
-          await this.reclaimTask();
-        }()
-      }.bind(this), nextClaim);
+        await queue.reportException(taskid, runId, {reason});
+
+        if (addArtifacts) {
+          let task = c.task;
+          // set the artifact expiration to match the task
+          let expiration = task.expires || taskcluster.fromNow(task.deadline, "1 year");
+          let content = {'taskId': primaryTaskId, 'runId': primaryRunId};
+          contentJson = JSON.Stringify(content);
+          await uploadToS3(c, contentJson, "public/superseded-by.json", expiration, {
+            'content-type': 'application/json',
+            'content-length': contentJson.length,
+          });
+
+          supersedes.push(content);
+        }
+      }));
+    } catch (e) {
+      // failing to resolve a non-primary claim is not a big deal: it will
+      // either time out and go back in the queue, or it was cancelled or
+      // otherwise modified while we were working on it.
+      log("while resolving superseded tasks: " + e);
+    }
+
+    if (addArtifacts) {
+      let task = this.claim.task;
+      let expiration = task.expires || taskcluster.fromNow(task.deadline, "1 year");
+      let contentJson = JSON.Stringify(supersedes)
+      await uploadToS3(c, contentJson, "public/supersedes.json", expiration, {
```

Comment:
Hrm, then I must be misreading.  Looking at it, it seems that it's taking a complete task handler, because then it looks for task.runtime.queue.
https://github.com/taskcluster/docker-worker/pull/193/files#diff-877d95d813ff49157dff0c2db4fcab51R5

Note: This is going to be redone soon because we do not want to use the global queue client, but rather create a new one using the temp creds given back by a claim. (wander is looking at adding this in q1).